### DISCUSSION TITLE
Replacing the last page in the stack

### DISCUSSION THREAD

**Discussion by osaxma** [csells/go_router#188] [Ideas]
**Use Case:**
Assume we've a page with nested navigation that can be pushed on top of the stack from anywhere in the app. 

Taking Twitter Profile page as an example, the user can navigate to any user profile from anywhere in the app such as:

```
/home -> /username1/tweet3 -> /username4/tweet24 -> /username 
```
If the user clicks on the `likes` tab of the profile, the stack would become:
```
/home -> /username1/tweet3 -> /username4/tweet24 -> /username/likes 
```

**Issue**:
Let's continue the same the same example with `go_router` and assume `/username` is a sub-route to `/home`. 

If the user clicks on a tab to change the nested navigation from `likes` to `media`, there are two options:

- Using `context.go('/username/media')` will reset the stack in the following manner (undesirable):
```
from:
/home -> /username1/tweet3 -> /username4/tweet24 -> /username/likes 
to:
/home -> /username/media 
```

- Using `context.push('/username/media')` will expand the stack in the following manner (also undesirable) : 

```
from:
/home -> /username1/tweet3 -> /username4/tweet24 -> /username/likes 
to:
/home -> /username1/tweet3 -> /username4/tweet24 -> /username/likes -> /username/media
```
The undesirable problem with `context.push` is that the back button of the `AppBar` will take the user back again to the profile. In addition, pushing the page is not UI friendly (video). I tried combining `pop` and `push` but the switch in UI was very obvious (same as push only) and could lead to state loss.

https://user-images.githubusercontent.com/46427323/143068182-237afbd5-8c0b-42c5-8c77-1578a837fc7f.mov

**Discussion**:
Looking at the public API of `go_router`, I could not find a way to replace the last page of the stack without the aforementioned effects.  

Is this possible and can it be supported?

**Comment by csells** [comment]
I understand the scenario -- you've described it well. Can you post a minimal repro using pop() + push()? I'm curious what UX you're getting in that case.

**Reply by osaxma** [reply]
I added an example here [my fork - nested_nav_deep branch](https://github.com/osaxma/go_router/blob/nested_nav_deep/example/lib/nested_nav_deep.dart)

**Comment by osaxma** [comment]
Hi @csells 

Just a follow up, in #193 , you mentioned the following:
> I've been thinking about how to produce a more general-purpose API for getting the current stack of pages and replacing it completely, which would make doing things like pushing, replacing, popping, etc. as easy as grabbing the current stack, rearranging it to your heart's content and then handing it back to go_router. Until that happens, it's premature to take PRs like this one.

I'm just wondering if this is planned anytime in the near future?

Thank you

**Comment by csells** [comment]
I've been noodling on it but haven't written any code yet. Stay tuned.
